### PR TITLE
Mvvm6

### PR DESCRIPTION
**Pull request by tbuschto**

### CODE REVIEW

File: examples/bind-one-way-jsx/README.md [hunk 1]
```diff
@@ -10,6 +10,6 @@ Demonstrates the use of the `bind` JSX attribute prefix to create one-way bindin
 
 The `ExampleComponent` property `myText` is bound to the `text` properties of multiple `TextView` children. The first two variants have the same effect, just using different syntax. The third variant demonstrates how to define a fallback value. It will be displayed when the property `myText` is set to `undefined`, which is the case when the checkbox is not checked.
 
-The other property `myObject` is of the type `Model`, which is defined in the same file as `ExampleComponent`. In the component the `Model` field `someString` is bound to the `TextView` property `text` and `someNumber` to the `ProgressBar` property `selection` properties, once with and without a fallback value.
+The other property `myObject` is of the type `Model`, which is defined in the same file as `ExampleComponent`. In the component the `Model` field `someNumber` is bound to the `ProgressBar` property `selection`. Nested in `Model` is another `OtherModel` whose property `someString` is bound to the last `TextView`. Since its `text` property has an initial text set, this value will be used as the fallback if the model is detached.
```
Comment (cpetrov) [inline examples/bind-one-way-jsx/README.md:13]: Is model detachment a well-known concept? Maybe add a documentation link?

File: test/subscribe.spec.ts [hunk 2]
```diff
@@ -331,4 +331,153 @@ describe('subscribe', () => {
...
+    let target: ModelA;
+    let sub: SinonSpy;
+    let cancel: () => void;
+
+    beforeEach(() => {
+      target = new ModelA();
+      sub = spy();
+    });
+
+    describe('to primitive type', () => {
```
Comment (cpetrov) [inline test/subscribe.spec.ts:498]: verb missing, a little vague - maybe "(when) subscribe(d) to primitive type changes"?

File: test/subscribe.spec.ts [hunk 3]
```diff
@@ -331,4 +331,153 @@ describe('subscribe', () => {
...
+
+        cancel();
+        target.foo = 2;
+
+        expect(sub).not.to.have.been.called;
+      });
+
+    });
+
+    describe('to nested target', () => {
```
Comment (cpetrov) [inline test/subscribe.spec.ts:541]: see previous comment

File: .vscode/launch.json [hunk 4]
```diff
@@ -14,7 +14,7 @@
       },
       "runtimeArgs": [
         "-p",
```
Comment (cpetrov) [inline .vscode/launch.json:16]: Maybe change commit summary to use "vscode" as "vs[space]code" does not seem common

File: examples/bind-one-way-jsx/README.md [hunk 5]
```diff
@@ -4,11 +4,11 @@
 
 ## Description
 
-This is the JavaScript/JSX variant of the [bind-one-way](../bind-one-way) TypeScript/JSX example.
+This is the JavaScript/JSX variant of the [bind-one-way](../bind-one-way) TypeScript/JSX example. Another TypeScript variant that uses only decorators without JSX can be found [here](../bind-one-way-ts).
 
-Demonstrates the use of the `bind` JSX attribute prefix to create one-way bindings from a custom component instance to its children. This app creates an instance of the included `ExampleComponent` class and a checkbox that allows to change the property values of that instance.
+Demonstrates the use of the `bind` JSX attribute prefix to create one-way bindings from a custom component instance to its children. This app creates an instance of the included `ExampleComponent` class and a checkboxes that allow to change the property values of that instance.
```
Comment (cpetrov) [inline examples/bind-one-way-jsx/README.md:9]: and a checkboxes -> and checkboxes

File: examples/bind-one-way-jsx/README.md [hunk 6]
```diff
@@ -4,11 +4,11 @@
...
 
-The `ExampleComponent` property `myText` is bound to the `text` properties of multiple `TextView` children. The first two variants have the same effect, just using different syntax. The third variant demonstrates how to define a fallback value. It will be displayed when the property `myText` is set to `undefined`, which is the case when the checkbox is not checked.
+The `ExampleComponent` property `myText` is bound to the `text` properties of a `TextView` child. It has a fallback value (the original `text` that will be displayed when the property `myText` is set to `undefined`. In this example that is the case when the first checkbox is not checked.
```
Comment (cpetrov) [inline examples/bind-one-way-jsx/README.md:11]: opening bracket does not have a matching closing bracket

File: examples/bind-one-way-ts/README.md [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+
+[![GitPod Logo](../../doc/run-in-gitpod.png)](https://gitpod.io/#example=bind-one-way/https://github.com/eclipsesource/tabris-decorators/tree/master/examples/bind-one-way)
+
+## Description
+
+Note: This TypeScript example uses no JSX syntax, only decorators. A JavaScript/JSX variant can be found [here](../bind-one-way-jsx), and a TypeScript/JSX variant [here](../bind-one-way).
+
+Demonstrates the use of the `@bind` decorator to create one-way bindings from a custom component instance to its children. This app creates an instance of the included `ExampleComponent` class and checkboxes that allow to change the property values of that instance.
+
+The `ExampleComponent` property `myText` is bound to the `text` properties of a `TextView` child. It has a fallback value (the original `text` that will be displayed when the property `myText` is set to `undefined`. In this example that is the case when the first checkbox is not checked.
```
Comment (cpetrov) [inline examples/bind-one-way-ts/README.md:11]: opening bracket does not have a matching closing bracket

File: examples/bind-one-way-ts/README.md [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+
+The other property `myObject` is of the type `Model`, which is defined in the same file as `ExampleComponent`. In the component the `Model` field `someNumber` is bound to the `ProgressBar` property `selection` by using a converter to extract `someNumber`.
+
+Nested in `Model` is another `OtherModel` whose property `someString` is bound to the last `TextView`. **This requires the `otherModel` property to be "observing"**, i.e. for its `@property` decorator to have `observe` set to `true`. Alternatively, the `@prop` decorator can be used. If `otherModel` wasn't observing its contained object, a change to `someString` would not be recognized. This applies to all properties of objects nested within the bound value. (However, the JSX data binding syntax does not have this requirement, only `@bind`/`@bindAll`.)
```
Comment (cpetrov) [inline examples/bind-one-way-ts/README.md:15]: idk if we shouldn't inline such docs in the example code? This would make the doc easier to maintain and make the association with the code easier to comprehend for the reader?

File: examples/bind-one-way-ts/src/ExampleComponent.ts [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+      this.set(properties);
+      this.append(
+        Stack({
+          spacing: 12,
+          padding: 12,
+          apply: Setter(TextView, {font: {size: 21}}),
+          children: [
+            TextView({text: 'Binding to a component property'}),
+            TextView({id: 'text1', text: 'Placeholder', background: 'yellow'}),
+            TextView({text: 'Binding to a object property:'}),
```
Comment (cpetrov) [inline examples/bind-one-way-ts/src/ExampleComponent.ts:36]: a -> an

File: examples/bind-one-way/README.md [hunk 10]
```diff
@@ -4,11 +4,11 @@
 
 ## Description
 
-Note: A JavaScript/JSX variant of the this TypeScript/JSX example can be found [here](../bind-one-way-jsx).
+Note: A JavaScript/JSX variant of the this TypeScript/JSX example can be found [here](../bind-one-way-jsx). Another TypeScript variant that uses only decorators without JSX can be found [here](../bind-one-way-ts).
 
-Demonstrates the use of the `bind` JSX attribute prefix to create one-way bindings from a custom component instance to its children. This app creates an instance of the included `ExampleComponent` class and a checkbox that allows to change the property values of that instance.
+Demonstrates the use of the `bind` JSX attribute prefix to create one-way bindings from a custom component instance to its children. This app creates an instance of the included `ExampleComponent` class and a checkboxes that allow to change the property values of that instance.
```
Comment (cpetrov) [inline examples/bind-one-way/README.md:9]: and a checkboxes -> and checkboxes

File: examples/bind-one-way/README.md [hunk 11]
```diff
@@ -4,11 +4,11 @@
...
 
-The `ExampleComponent` property `myText` is bound to the `text` properties of multiple `TextView` children. The first two variants have the same effect, just using different syntax. The third variant demonstrates how to define a fallback value. It will be displayed when the property `myText` is set to `undefined`, which is the case when the checkbox is not checked.
+The `ExampleComponent` property `myText` is bound to the `text` properties of a `TextView` child. It has a fallback value (the original `text` that will be displayed when the property `myText` is set to `undefined`. In this example that is the case when the first checkbox is not checked.
```
Comment (cpetrov) [inline examples/bind-one-way/README.md:11]: missing closing bracket

File: examples/property-change-events-jsx/README.md [hunk 12]
```diff
@@ -6,9 +6,11 @@
...
 
-Both `Person` properties are decorated with `@property` and have a matching `ChangeListeners` field. E.g. `age` is a property of the type `number` and `onAgeChanged` is of the type `ChangeListeners<Person, 'age'>`, indicating that it fires change events for the `age` property of `Person`. When the value of `age` is changed `@property` looks for `onAgeChanged` by name (`on`, followed by the upper case property name, followed by `Changed`) and triggers a matching change event.
+The `Person` properties `age` and `name` are decorated with `@property` and have matching `ChangeListeners` fields. E.g. `age` is a property of the type `number` and `onAgeChanged` is of the type `ChangeListeners<Person, 'age'>`, indicating that it fires change events for the `age` property of `Person`. When the value of `age` is changed `@property` looks for `onAgeChanged` by name (`on`, followed by the upper case property name, followed by `Changed`) and triggers a matching change event.
```
Comment (cpetrov) [inline examples/property-change-events-jsx/README.md:11]: When the value of `age` is changed[,]

File: examples/property-change-events/README.md [hunk 13]
```diff
@@ -6,9 +6,11 @@
...
 
-Both `Person` properties are decorated with `@property` and have a matching `ChangeListeners` field. E.g. `age` is a property of the type `number` and `onAgeChanged` is of the type `ChangeListeners<Person, 'age'>`, indicating that it fires change events for the `age` property of `Person`. When the value of `age` is changed `@property` looks for `onAgeChanged` by name (`on`, followed by the upper case property name, followed by `Changed`) and triggers a matching change event.
+The `Person` properties `age` and `name` are decorated with `@property` and have matching `ChangeListeners` fields. E.g. `age` is a property of the type `number` and `onAgeChanged` is of the type `ChangeListeners<Person, 'age'>`, indicating that it fires change events for the `age` property of `Person`. When the value of `age` is changed `@property` looks for `onAgeChanged` by name (`on`, followed by the upper case property name, followed by `Changed`) and triggers a matching change event.
```
Comment (cpetrov) [inline examples/property-change-events/README.md:11]: When the value of `age` is changed[,]

File: src/decorators/bind.ts [hunk 14]
```diff
@@ -122,7 +122,8 @@ export function bind(...args: any[]): any {
       equals: isShorthand ? null : args[0].equals || null,
       convert: getConverter(isShorthand, args),
       nullable: isShorthand ? null : 'nullable' in args[0] ? args[0].nullable : null,
-      default: isShorthand ? undefined : args[0].default
+      default: isShorthand ? undefined : args[0].default,
+      observe: isShorthand ? true : args[0].observe !== false
```
Comment (cpetrov) [inline src/decorators/bind.ts:126]: `isShorthand || args[0].observe !== false`?

File: src/internals/CustomPropertyDescriptor.ts [hunk 15]
```diff
@@ -134,9 +147,21 @@ export class CustomPropertyDescriptor<Proto extends object, TargetType> {
       if (!this.isUnchecked()) {
         this.checkType(newValue);
       }
-      getPropertyStore(instance).set(this.propertyName, newValue);
+      const store = getPropertyStore(instance);
+      store.set(this.propertyName, newValue);
       this.changeHandler.forEach(handler => handler(instance, newValue, oldValue));
-      trigger(instance, this.changeEvent, {value: newValue});
+      if (this.observe) {
```
Comment (cpetrov) [inline src/internals/CustomPropertyDescriptor.ts:153]: reverse if condition to bring `if` and `else` closer together

File: test/component-bind.spec.ts [hunk 16]
```diff
@@ -297,9 +299,16 @@ describe('component', () => {
     describe('with ">>" prefix', () => {
 
       @component class ToRight extends Composite {
+
         // space after direction is not required
```
Comment (cpetrov) [inline test/component-bind.spec.ts:303]: No need to demonstrate this edge case, as this sounds like the expected behavior?

File: test/subscribe.spec.ts [hunk 17]
```diff
@@ -331,6 +331,149 @@ describe('subscribe', () => {
...
+    let target: ModelA;
+    let sub: SinonSpy;
+    let cancel: () => void;
+
+    beforeEach(() => {
+      target = new ModelA();
+      sub = spy();
+    });
+
+    describe('to primitive type', () => {
```
Comment (cpetrov) [inline test/subscribe.spec.ts:356]: had related comments in another similar spec

**Review by cpetrov — CHANGES_REQUESTED** [review]

File: examples/bind-one-way-ts/README.md (hunk 8 above)
Comment (tbuschto) [reply examples/bind-one-way-ts/README.md:15]: Good idea, but not now.

File: test/component-bind.spec.ts (hunk 16 above)
Comment (tbuschto) [reply test/component-bind.spec.ts:303]: ??

File: src/decorators/bind.ts (hunk 14 above)
Comment (tbuschto) [reply src/decorators/bind.ts:126]: I find mine more expressive
